compiler/sem/semexprs.nim

```diff
@@ -1867,26 +1871,48 @@ proc takeImplicitAddr(c: PContext, n: PNode; isLent: bool): PNode =
     # `proc fun(a: var int): var int = a`
     discard
   else: discard
-  let valid = isAssignable(c, n, isLent)
-  if valid != arLValue:
-    if valid == arLocalLValue:
-      localReport(c.config, n, reportSem rsemLocalEscapesStackFrame)
+  
+  case isAssignable(c, n, isLent)
+  of arLValue:
+    discard
+  of arLocalLValue:
+    localReport(c.config, n, reportSem rsemLocalEscapesStackFrame)
+  else:
+    localReport(c.config, n, reportSem rsemExprHasNoAddress)
+
+  let typ =
+    case n.typ.kind
+    of {tyVar, tyLent}:
+      n.typ
     else:
-      localReport(c.config, n, reportSem rsemExprHasNoAddress)
-  result = newNodeIT(nkHiddenAddr, n.info, if n.typ.kind in {tyVar, tyLent}: n.typ else: makePtrType(c, n.typ))
+      makePtrType(c, n.typ)
+  result = newNodeIT(nkHiddenAddr, n.info, typ)
   result.add(n)
 
 proc asgnToResultVar(c: PContext, n, le, ri: PNode) {.inline.} =
-  if le.kind == nkHiddenDeref:
+  case le.kind
+  of nkHiddenDeref:
     var x = le[0]
-    if x.kind == nkSym:
-      if x.sym.kind == skResult and (x.typ.kind in {tyVar, tyLent} or classifyViewType(x.typ) != noView):
-        n[0] = x # 'result[]' --> 'result'
-        n[1] = takeImplicitAddr(c, ri, x.typ.kind == tyLent)
-        x.typ.flags.incl tfVarIsPtr
-        #echo x.info, " setting it for this type ", typeToString(x.typ), " ", n.info
-      elif sfGlobal in x.sym.flags:
-        x.typ.flags.incl tfVarIsPtr
+    case x.kind
+    of nkSym:
+      case x.sym.kind
+      of skResult:
+        let varLentOrView = x.typ.kind in {tyVar, tyLent} or
+                            classifyViewType(x.typ) != noView
+        
+        if x.sym.kind == skResult and varLentOrView:
+          n[0] = x # 'result[]' --> 'result'
+          n[1] = takeImplicitAddr(c, ri, x.typ.kind == tyLent)
+          x.typ.flags.incl tfVarIsPtr
+          #echo x.info, " setting it for this type ", typeToString(x.typ), " ", n.info
+        elif sfGlobal in x.sym.flags:
+          x.typ.flags.incl tfVarIsPtr
+      else:
+        discard
```

Comment:
This is not equivalent to the original logic, if you have `x.sym.kind != skResult and sfGlobal in x.sym.flags` the old logic would still do `x.typ.flags.incl tfVarIsPtr`, the new one would lead to `discard`.
In this case I think the old version is more readable.

Reply:
Boo old style, this is tired me making an obvious error, let me see what another go might like like.

Interesting how no tests failed...

Reply:
had another go at it, bit cleaner, but I still think this code and its collaborators need a revisit.